bindings/matrix-sdk-ffi/src/widget.rs

```diff
@@ -0,0 +1,154 @@
+use std::sync::Arc;
+
+use matrix_sdk::async_trait;
+
+use crate::room::Room;
+
+#[derive(uniffi::Record)]
+pub struct Widget {
+    /// Information about the widget.
+    pub info: WidgetInfo,
+    /// Communication channels with a widget.
+    pub comm: Arc<WidgetComm>,
+}
+
+impl From<Widget> for matrix_sdk::widget::Widget {
+    fn from(value: Widget) -> Self {
+        let comm = &value.comm.0;
+        Self {
+            info: value.info.into(),
+            comm: matrix_sdk::widget::Comm { from: comm.from.clone(), to: comm.to.clone() },
+        }
+    }
+}
+
+/// Information about a widget.
+#[derive(uniffi::Record)]
+pub struct WidgetInfo {
+    /// Widget's unique identifier.
+    pub id: String,
+    /// Whether or not the widget should be initialized on load message
+    /// (`ContentLoad` message), or upon creation/attaching of the widget to
+    /// the SDK's state machine that drives the API.
+    pub init_on_load: bool,
+}
+
+impl From<WidgetInfo> for matrix_sdk::widget::Info {
+    fn from(value: WidgetInfo) -> Self {
+        let WidgetInfo { id, init_on_load } = value;
+        Self { id, init_on_load }
+    }
+}
+
+/// Communication "pipes" with a widget.
+#[derive(uniffi::Object)]
+pub struct WidgetComm(matrix_sdk::widget::Comm);
+
+/// Permissions that a widget can request from a client.
+#[derive(uniffi::Record)]
+pub struct Permissions {
+    /// Types of the messages that a widget wants to be able to fetch.
+    pub read: Vec<EventFilter>,
+    /// Types of the messages that a widget wants to be able to send.
+    pub send: Vec<EventFilter>,
+}
+
+impl From<Permissions> for matrix_sdk::widget::Permissions {
+    fn from(value: Permissions) -> Self {
+        Self {
+            read: value.read.into_iter().map(Into::into).collect(),
+            send: value.send.into_iter().map(Into::into).collect(),
+        }
+    }
+}
+
+impl From<matrix_sdk::widget::Permissions> for Permissions {
+    fn from(value: matrix_sdk::widget::Permissions) -> Self {
+        Self {
+            read: value.read.into_iter().map(Into::into).collect(),
+            send: value.send.into_iter().map(Into::into).collect(),
+        }
+    }
+}
+
+/// Different kinds of filters that could be applied to the timeline events.
+#[derive(uniffi::Enum)]
+pub enum EventFilter {
```

Comment:
Is it specific to the widget? Then can we rename it `WidgetEventFilter`? Otherwise, let's move it inside the `timeline` FFI module?